metron-contrib/metron-performance/src/main/java/org/apache/metron/performance/sampler/BiasedSampler.java

```diff
@@ -0,0 +1,95 @@
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.TreeMap;
+
+public class BiasedSampler implements Sampler {
+  TreeMap<Double, Map.Entry<Integer, Integer>> discreteDistribution;
+  public BiasedSampler(List<Map.Entry<Integer, Integer>>  discreteDistribution, int max) {
+    this.discreteDistribution = createDistribution(discreteDistribution, max);
+  }
+
+  public static List<Map.Entry<Integer, Integer>> readDistribution(File distrFile) throws IOException {
+    List<Map.Entry<Integer, Integer>> ret = new ArrayList<>();
+    System.out.println("Using biased sampler with the following biases:");
+    try(BufferedReader br = new BufferedReader(new FileReader(distrFile))) {
+      int sumLeft = 0;
+      int sumRight = 0;
+      for(String line = null;(line = br.readLine()) != null;) {
+        if(line.startsWith("#")) {
+          continue;
+        }
+        Iterable<String> it = Splitter.on(",").split(line.trim());
+        int left = Integer.parseInt(Iterables.getFirst(it, null));
+        int right = Integer.parseInt(Iterables.getLast(it, null));
+        System.out.println("\t" + left + "% of templates will comprise roughly " + right + "% of sample output");
+        ret.add(new AbstractMap.SimpleEntry<>(left, right));
+        sumLeft += left;
+        sumRight += right;
+      }
+      if(sumLeft > 100 || sumRight > 100 ) {
+        throw new IllegalStateException("Neither columns must sum to beyond 100.  " +
+                "The first column is the % of templates. " +
+                "The second column is the % of the sample that % of template occupies.");
+      }
+      else if(sumLeft < 100 && sumRight < 100) {
+        int left = 100 - sumLeft;
+        int right = 100 - sumRight;
+        System.out.println("\t" + left + "% of templates will comprise roughly " + right + "% of sample output");
+        ret.add(new AbstractMap.SimpleEntry<>(left, right));
+      }
+      return ret;
+    }
+  }
+
+  private static TreeMap<Double, Map.Entry<Integer, Integer>>
+                 createDistribution(List<Map.Entry<Integer, Integer>>  discreteDistribution, int max) {
+    TreeMap<Double, Map.Entry<Integer, Integer>> ret = new TreeMap<>();
+    int from = 0;
+    double weight = 0.0d;
+    for(Map.Entry<Integer, Integer> kv : discreteDistribution) {
+      double pctVals = kv.getKey()/100.0;
+      int to = from + (int)(max*pctVals);
+      double pctWeight = kv.getValue()/100.0;
+      ret.put(weight, new AbstractMap.SimpleEntry<>(from, to));
+      weight += pctWeight;
+      from = to;
+    }
+    return ret;
+  }
+
+  @Override
+  public int sample(Random rng, int limit) {
+    double weight = rng.nextDouble();
+    Map.Entry<Integer, Integer> range = discreteDistribution.floorEntry(weight).getValue();
+    return rng.nextInt(range.getValue() - range.getKey()) + range.getKey();
```

Comment:
Jerks like me who provide negatives or zeroes cause ugly exceptions to start showing up. because the nextInt doesn't get happy values.

Reply:
done